enterprise/wsproxy/keycache_test.go

```diff
@@ -0,0 +1,504 @@
+
+		// Assert hits cache.
+		got, err = cache.Signing(ctx)
+		require.NoError(t, err)
+		require.Equal(t, newKey, got)
+		require.Equal(t, 2, fc.called)
+
+		// We check again to ensure the timer has been reset.
+		_, advance = clock.AdvanceNext()
+		advance.MustWait(ctx)
+		require.Equal(t, 3, fc.called)
+		require.Equal(t, time.Minute*10, dur)
+	})
+
+	// This test ensures that if the refresh timer races with an inflight request
+	// and loses that it doesn't cause a redundant fetch.
+
+	t.Run("RefreshNoDoubleFetch", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx    = testutil.Context(t, testutil.WaitShort)
+			logger = slogtest.Make(t, nil)
+			clock  = quartz.NewMock(t)
+		)
+
+		now := clock.Now().UTC()
+		expected := codersdk.CryptoKey{
+			Feature:   codersdk.CryptoKeyFeatureWorkspaceApp,
+			Secret:    "key1",
+			Sequence:  12,
+			StartsAt:  now,
+			DeletesAt: now.Add(time.Minute * 10),
+		}
+		fc := newFakeCoderd(t, []codersdk.CryptoKey{
+			expected,
+		})
+
+		// Create a trap that blocks when the refresh timer fires.
+		trap := clock.Trap().Now("refresh")
+		cache, err := wsproxy.NewCryptoKeyCache(ctx, logger, wsproxysdk.New(fc.url), withClock(clock))
+		require.NoError(t, err)
+
+		_, wait := clock.AdvanceNext()
+		trapped := trap.MustWait(ctx)
+
+		newKey := codersdk.CryptoKey{
+			Feature:  codersdk.CryptoKeyFeatureWorkspaceApp,
+			Secret:   "key2",
+			Sequence: 13,
+			StartsAt: now,
+		}
+		fc.keys = []codersdk.CryptoKey{newKey}
+
+		_, err = cache.Verifying(ctx, newKey.Sequence)
+		require.NoError(t, err)
+		require.Equal(t, 2, fc.called)
+
+		trapped.Release()
+		wait.MustWait(ctx)
+		require.Equal(t, 2, fc.called)
+		trap.Close()
+
+		// The next timer should fire in 10 minutes.
+		dur, wait := clock.AdvanceNext()
+		wait.MustWait(ctx)
+		require.Equal(t, time.Minute*10, dur)
+		require.Equal(t, 3, fc.called)
+	})
+
+	t.Run("Closed", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx    = testutil.Context(t, testutil.WaitShort)
+			logger = slogtest.Make(t, nil)
+			clock  = quartz.NewMock(t)
+		)
+
+		now := clock.Now()
+		expected := codersdk.CryptoKey{
+			Feature:  codersdk.CryptoKeyFeatureWorkspaceApp,
+			Secret:   "key1",
+			Sequence: 12,
+			StartsAt: now,
+		}
+		fc := newFakeCoderd(t, []codersdk.CryptoKey{
+			expected,
+		})
+
+		cache, err := wsproxy.NewCryptoKeyCache(ctx, logger, wsproxysdk.New(fc.url), withClock(clock))
+		require.NoError(t, err)
+
+		got, err := cache.Signing(ctx)
+		require.NoError(t, err)
+		require.Equal(t, expected, got)
+		require.Equal(t, 1, fc.called)
+
+		got, err = cache.Verifying(ctx, expected.Sequence)
+		require.NoError(t, err)
+		require.Equal(t, expected, got)
+		require.Equal(t, 1, fc.called)
+
+		cache.Close()
+
+		_, err = cache.Signing(ctx)
+		require.ErrorIs(t, err, cryptokeys.ErrClosed)
+
+		_, err = cache.Verifying(ctx, expected.Sequence)
+		require.ErrorIs(t, err, cryptokeys.ErrClosed)
+	})
+}
+
+type fakeCoderd struct {
+	server *httptest.Server
```

Comment:
If you made `CryptoKeyCache.client` an interface, you wouldn't need to mess around with HTTP, URLs, or JSON.  You could just directly return the keys you want to test with.